modules/currency-l1/src/main/scala/io/constellationnetwork/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,55 +47,103 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+                      .map(EstimatedFeeResponse(estimatedFee, _))
+                  }
+                  .map(_.asJson.dropNullValues)
+                  .flatMap(Ok(_))
+            }
           }
         }
         .handleUnknownError
 
     case GET -> Root / "data" =>
-      implicit val signedEncoder: Encoder[Signed[DataUpdate]] = Signed.encoder(dataApplication.dataEncoder)
-      dataUpdatesQueue.view.flatMap(Ok(_))
+      implicit val dataUpdateEncoder: Encoder[DataUpdate] = dataApplication.dataEncoder
+      dataTransactionsQueue.view.flatMap(Ok(_))
 
     case req @ POST -> Root / "data" =>
-      implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def handleDataUpdates(dataResponse: DataResponse): F[Response[F]] = {
+        val dataTransactions = dataResponse match {
+          case SingleDataUpdateResponse(dataUpdate)   => NonEmptyList.one(dataUpdate)
+          case DataTransactionsResponse(transactions) => transactions
+        }
+
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .map(_.leftMap(_.widen[DataTransaction]))
+                  .map(_.widen[Hashed[DataTransaction]])
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .map(_.leftMap(_.widen[DataTransaction]))
+                  .map(_.widen[Hashed[DataTransaction]])
+            }
+          }.flatMap { results =>
+            val (errors, validTransactions) = results.toList.partitionMap(identity)
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
+              )
+            } else {
+              NonEmptyList.fromList(validTransactions) match {
+                case None =>
+                  BadRequest(Json.obj("error" -> Json.fromString("No valid data transactions found")))
+                case Some(txs) =>
+                  dataTransactionsQueue
+                    .offer(txs.map(_.signed))
+                    .flatMap(_ => dataApplication.responseDecoder(dataResponse, txs))
               }
```

Comment:
Ok, so we indeed terminate the processing early. I think we should:
1. get validTransactions and enqueue them
2. either put `results` to `dataApplication.responseEncoder` or `valid, invalid` as separate properties, so developer can react on errors whatever he likes to. all other/unknown errors should be propagated through the effect type `F`

Reply:
👍🏼